testify/src/main/java/testify/annotation/logging/LogRecorder.java

```diff
@@ -0,0 +1,370 @@
+        // must only be called within the parent process
+        final LogFormatter formatter = new LogFormatter(startTime, partNameWidth);
+        dispatchRequestAndWaitForReply(dedicatedBus, REQUEST_LOG_RECORDS, formatter);
+        // and then retrieve the result
+        return dedicatedBus.get(REPLY_LOG_RECORDS);
+    }
+
+    static class LogFormatter implements Function<LogRecord, String> {
+        enum Spec implements TypeSpec<LogFormatter> {
+            REQUEST_LOG_RECORDS;
+            public String stringify(LogFormatter f) { return f.startTime + " " + f.partNameWidth; }
+            public LogFormatter unstringify(String s) {
+                Scanner scan = new Scanner(s);
+                return new LogFormatter(scan.nextLong(), scan.nextInt());
+            }
+        }
+        final long startTime;
+        final int partNameWidth;
+        transient final String format;
+        transient LogRecorder recorder;
+
+        LogFormatter(long startTime, int partNameWidth) {
+            this.startTime = startTime;
+            this.partNameWidth = partNameWidth;
+            this.format = "LOG: %02d.%03d  %" + partNameWidth + "s %8s  [%s] %7s: %s";
+        }
+
+        @Override
+        public String apply(LogRecord rec) {
+            long millis = rec.getMillis() - startTime;
+            String result = String.format(format,
+                    millis / 1000,
+                    millis % 1000,
+                    recorder.processName,
+                    recorder.threadNames.get((long) rec.getThreadID()),
+                    rec.getLoggerName(),
+                    rec.getLevel(),
+                    rec.getMessage());
+            Throwable throwable = rec.getThrown();
+            if (null != throwable) result = String.format("%s%n%s", result, formatThrowable(throwable));
+            return result;
+        }
+
+        @Override
+        public String toString() {
+            return "LogFormatter { startTime=" + startTime + ", format=" + format + "}";
+        }
+    }
+
+    private synchronized void replyLogRecords(LogFormatter formatter) {
+        formatter.recorder = this;
+        List<String> logRecords = drainInOrder(journals).map(formatter).collect(toList());
+        dedicatedBus.put(REPLY_LOG_RECORDS, logRecords);
+        dispatchReply("returned log records");
+    }
+
+    private static String formatThrowable(Throwable t) {
+        try (StringWriter sw = new StringWriter()) {
+            try (PrintWriter out = new PrintWriter(sw)) {
+                out.printf("Exception was %s%n", t);
+                t.printStackTrace(out);
+                return sw.toString();
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    static void close(Bus dedicatedBus) { dispatchRequestAndWaitForReply(dedicatedBus, CLOSE); }
+
+    private synchronized void close() {
```

Comment:
Is this close() idempotent ?
Should LogRecorder declare itself as AutoCloseable ?

Reply:
I tried AutoCloseable, but IntelliJ complained when it was created outside of a try-with-resources statement.